File: plugins/haskell-lsp/tests/SampleTest.hs
-- Hook test
{-# LANGUAGE OverloadedStrings #-}

-- | Sample Haskell module for LSP plugin validation.
--
-- This file contains various Haskell constructs to test:
-- - LSP operations (hover, go to definition, type information)
-- - Hook validation (linting, formatting, type checking)
module SampleTest where

import Data.List (find)
import Data.Maybe (fromMaybe, isJust)
import Test.HUnit

-- | Represents a user in the system.
data User = User
  { userName :: String,
    userEmail :: String,
    userAge :: Int
  }
  deriving (Show, Eq)

-- | Creates a greeting message for the user.
greet :: User -> String
greet user = "Hello, " ++ userName user ++ "!"

-- | Checks if the user is an adult (18+).
isAdult :: User -> Bool
isAdult user = userAge user >= 18

-- | Service for managing users.
data UserService = UserService
  { users :: [User]
  }
  deriving (Show)

-- | Creates a new empty UserService.
newUserService :: UserService
newUserService = UserService {users = []}

-- | Adds a user to the service.
addUser :: User -> UserService -> UserService
addUser user service = service {users = user : users service}

-- | Finds a user by email address.
findByEmail :: String -> UserService -> Maybe User
findByEmail email service = find (\u -> userEmail u == email) (users service)

-- | Returns the number of users in the service.
count :: UserService -> Int
count = length . users

-- | Calculates the average of a list of numbers.
-- Returns Nothing if the list is empty.
calculateAverage :: [Double] -> Maybe Double
calculateAverage [] = Nothing
calculateAverage xs = Just (sum xs / fromIntegral (length xs))

-- | Safe division that returns Nothing for division by zero.
safeDivide :: Double -> Double -> Maybe Double
safeDivide _ 0 = Nothing
safeDivide x y = Just (x / y)

-- | Filters adult users from a list.
filterAdults :: [User] -> [User]
filterAdults = filter isAdult

-- | Gets the average age of users, or 0 if no users.
averageAge :: [User] -> Double
averageAge [] = 0
averageAge us = fromMaybe 0 (calculateAverage ages)
  where
    ages = map (fromIntegral . userAge) us

-- TODO: Add more comprehensive user validation
-- FIXME: Handle email validation properly

-- Test cases using HUnit

testUserGreet :: Test
testUserGreet =
  TestCase $
    assertEqual
      "greet should return proper greeting"
      "Hello, Alice!"
      (greet alice)
  where
    alice = User {userName = "Alice", userEmail = "alice@example.com", userAge = 25}

testUserIsAdult :: Test
testUserIsAdult =
  TestList
    [ TestCase $ assertEqual "25 year old is adult" True (isAdult adult),
      TestCase $ assertEqual "15 year old is not adult" False (isAdult minor),
      TestCase $ assertEqual "18 year old is adult" True (isAdult exact)
    ]
  where
    adult = User {userName = "Adult", userEmail = "adult@example.com", userAge = 25}
    minor = User {userName = "Minor", userEmail = "minor@example.com", userAge = 15}
    exact = User {userName = "Exact", userEmail = "exact@example.com", userAge = 18}

testUserService :: Test
testUserService =
  TestList
    [ TestCase $ assertEqual "initial count is 0" 0 (count service),
      TestCase $ assertEqual "count after adding user" 1 (count service'),
      TestCase $ assertEqual "findByEmail finds user" True (isJust found),
      TestCase $ assertEqual "found user has correct name" "Bob" (maybe "" userName found)
    ]
  where
    service = newUserService
    bob = User {userName = "Bob", userEmail = "bob@example.com", userAge = 30}
    service' = addUser bob service
    found = findByEmail "bob@example.com" service'

testCalculateAverage :: Test
testCalculateAverage =
  TestList
    [ TestCase $ assertEqual "average of [1,2,3,4,5]" (Just 3.0) (calculateAverage [1, 2, 3, 4, 5]),
      TestCase $ assertEqual "average of empty list" Nothing (calculateAverage []),
      TestCase $ assertEqual "average of single element" (Just 42.0) (calculateAverage [42.0])
    ]

testSafeDivide :: Test
testSafeDivide =
  TestList
    [ TestCase $ assertEqual "10 / 2 = 5" (Just 5.0) (safeDivide 10 2),
      TestCase $ assertEqual "division by zero" Nothing (safeDivide 10 0),
      TestCase $ assertEqual "0 / 5 = 0" (Just 0.0) (safeDivide 0 5)
    ]

testFilterAdults :: Test
testFilterAdults =
  TestCase $ assertEqual "filter adults from mixed list" 2 (length (filterAdults mixedUsers))
  where
    mixedUsers =
      [ User {userName = "Adult1", userEmail = "a1@example.com", userAge = 25},
        User {userName = "Minor1", userEmail = "m1@example.com", userAge = 15},
        User {userName = "Adult2", userEmail = "a2@example.com", userAge = 30},
        User {userName = "Minor2", userEmail = "m2@example.com", userAge = 12}
      ]

testAverageAge :: Test
testAverageAge =
  TestList
    [ TestCase $ assertEqual "average age of users" 25.0 (averageAge testUsers),
      TestCase $ assertEqual "average age of empty list" 0.0 (averageAge [])
    ]
  where
    testUsers =
      [ User {userName = "User1", userEmail = "u1@example.com", userAge = 20},
        User {userName = "User2", userEmail = "u2@example.com", userAge = 30}
      ]

-- | Main test suite
tests :: Test
tests =
  TestList
    [ TestLabel "User Greet" testUserGreet,
      TestLabel "User IsAdult" testUserIsAdult,
      TestLabel "UserService" testUserService,
      TestLabel "Calculate Average" testCalculateAverage,
      TestLabel "Safe Divide" testSafeDivide,
      TestLabel "Filter Adults" testFilterAdults,
      TestLabel "Average Age" testAverageAge
    ]

-- | Run all tests
main :: IO ()
main = do
  _ <- runTestTT tests
  return ()

File: plugins/go-lsp/tests/sample_test.go
// Hook test
// Package main provides sample Go code for LSP plugin validation.
//
// This file contains various Go constructs to test:
// - LSP operations (hover, go to definition, references)
// - Hook validation (linting, formatting, testing)
package main

import (
	"errors"
	"testing"
)

// User represents a user in the system.
type User struct {
	Name  string
	Email string
	Age   int
}

// Greet returns a greeting message for the user.
func (u *User) Greet() string {
	return "Hello, " + u.Name + "!"
}

// IsAdult checks if the user is an adult (18+).
func (u *User) IsAdult() bool {
	return u.Age >= 18
}

// UserService manages users.
type UserService struct {
	users []User
}

// NewUserService creates a new UserService.
func NewUserService() *UserService {
	return &UserService{
		users: make([]User, 0),
	}
}

// AddUser adds a user to the service.
func (s *UserService) AddUser(user User) {
	s.users = append(s.users, user)
}

// FindByEmail finds a user by email.
func (s *UserService) FindByEmail(email string) (*User, bool) {
	for i := range s.users {
		if s.users[i].Email == email {
			return &s.users[i], true
		}
	}
	return nil, false
}

// Count returns the number of users.
func (s *UserService) Count() int {
	return len(s.users)
}

// CalculateAverage calculates the average of a slice of numbers.
func CalculateAverage(numbers []float64) (float64, error) {
	if len(numbers) == 0 {
		return 0, errors.New("cannot calculate average of empty slice")
	}

	var sum float64
	for _, n := range numbers {
		sum += n
	}
	return sum / float64(len(numbers)), nil
}

// TODO: Add more test cases
// FIXME: Handle edge cases

func TestUserGreet(t *testing.T) {
	user := User{Name: "Alice", Email: "alice@example.com", Age: 25}
	expected := "Hello, Alice!"
	if got := user.Greet(); got != expected {
		t.Errorf("Greet() = %q, want %q", got, expected)
	}
}

func TestUserIsAdult(t *testing.T) {
	tests := []struct {
		name     string
		age      int
		expected bool
	}{
		{"adult", 25, true},
		{"minor", 15, false},
		{"exact", 18, true},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			user := User{Name: "Test", Email: "test@example.com", Age: tt.age}
			if got := user.IsAdult(); got != tt.expected {
				t.Errorf("IsAdult() = %v, want %v", got, tt.expected)
			}
		})
	}
}

func TestUserService(t *testing.T) {
	service := NewUserService()

	user := User{Name: "Bob", Email: "bob@example.com", Age: 30}
	service.AddUser(user)

	if service.Count() != 1 {
		t.Errorf("Count() = %d, want 1", service.Count())
	}

	found, ok := service.FindByEmail("bob@example.com")
	if !ok {
		t.Error("FindByEmail() did not find user")
	}
	if found.Name != "Bob" {
		t.Errorf("found.Name = %q, want %q", found.Name, "Bob")
	}
}

func TestCalculateAverage(t *testing.T) {
	t.Run("positive numbers", func(t *testing.T) {
		avg, err := CalculateAverage([]float64{1, 2, 3, 4, 5})
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if avg != 3 {
			t.Errorf("CalculateAverage() = %f, want 3", avg)
		}
	})

	t.Run("empty slice", func(t *testing.T) {
		_, err := CalculateAverage([]float64{})
		if err == nil {
			t.Error("expected error for empty slice")
		}
	})
}

func BenchmarkCalculateAverage(b *testing.B) {
	numbers := make([]float64, 1000)
	for i := range numbers {
		numbers[i] = float64(i)
	}

	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		CalculateAverage(numbers)
	}
}
